src/main/java/com/github/kagkarlsson/scheduler/Scheduler.java

```diff
@@ -237,6 +239,10 @@ private void releaseExecutor(Execution execution) {
 			LOG.error("Released execution was not found in collection of executions currently being processed. Should never happen.");
 			statsRegistry.registerUnexpectedError();
 		}
+
+		if(pollAfterCompletion && currentlyProcessing.isEmpty()) {
```

Comment:
This should work for now. 

Am thinking that ideally we need to re-write the way `executeDue` works, perhaps to a model where the threads poll blockingly and continously from a queue of executions, and some other thread populates it. But that is a lot more work..

Reply:
Just a small formatting issue ;)

Reply:
Is `pollAfterLastCompleted` slightly better naming-wise?

Reply:
Yeah, wasn't 100% sure about the name yet. Your suggestion is a better name.

I also agree that ```executeDue``` could be more intelligent; Although your suggestion is a nice way to design it (one thread polls database and pushes tasks to the local execute queue, producer/consumer model), it may be overkill.

And in the end, the code for the producer would essentially be the same as in ```executeDue```.

Something like this?
```
jobsDue = fetch_due_jobs();

while(jobsDue.size() > 0 && we_need_to_keep_running) {
   try_to_pick_jobs(jobsDue);

   if(local_queue_full) {
      wait_for_queue_to_have_space();
   }

   jobsDue = fetch_due_jobs();
}

// exit loop, fall back to poll interval which triggers above loop again
```

Also something to ponder: when one spot in the thread pool opens up, do you fetch, pick and run a task, or do you batch them (50% space in thread pool, 2 threads, ...), or do you wait for the thread pool to be completely done (as in my current code)?

Reply:
> Also something to ponder: when one spot in the thread pool opens up, do you fetch, pick and run a task, or do you batch them (50% space in thread pool, 2 threads, ...), or do you wait for the thread pool to be completely done (as in my current code)?

Yeah this is also an aspect of the suggested solution. If you have 9 fast executions and 1 slow, the next batch will not be started after the slow one, so 9 threads will be more idle than necessary. A producer/consumer model will make that easier to fix.. Though I see what you are hinting at in your pseudo-code, hmm..

Reply:
Ideally you run as soon as there is a thread available

Reply:
I think that with a sensible default polling limit it makes (more) sense to run whenever a thread is free. This will basically prevent the "producer" from sleeping much while there are more tasks that are due.

Reply:
> Just a small formatting issue ;)

Sorry, I'm not sure what you meant here, can you explain?

Reply:
Just a cosmetic thing, missing a whitespace `if (`  ;)